notify/receivers.go

```diff
@@ -292,3 +316,189 @@ func ProcessNotifierError(config *GrafanaReceiver, err error) error {
 
 	return err
 }
+
+// GrafanaReceiverTyped represents a parsed and validated APIReceiver
+type GrafanaReceiverTyped struct {
+	Name                string
+	AlertmanagerConfigs []*NotifierConfig[alertmanager.Config]
+	DingdingConfigs     []*NotifierConfig[dinding.Config]
+	DiscordConfigs      []*NotifierConfig[discord.Config]
+	EmailConfigs        []*NotifierConfig[email.Config]
+	GooglechatConfigs   []*NotifierConfig[googlechat.Config]
+	KafkaConfigs        []*NotifierConfig[kafka.Config]
+	LineConfigs         []*NotifierConfig[line.Config]
+	OpsgenieConfigs     []*NotifierConfig[opsgenie.Config]
+	PagerdutyConfigs    []*NotifierConfig[pagerduty.Config]
+	PushoverConfigs     []*NotifierConfig[pushover.Config]
+	SensugoConfigs      []*NotifierConfig[sensugo.Config]
+	SlackConfigs        []*NotifierConfig[slack.Config]
+	TeamsConfigs        []*NotifierConfig[teams.Config]
+	TelegramConfigs     []*NotifierConfig[telegram.Config]
+	ThreemaConfigs      []*NotifierConfig[threema.Config]
+	VictoropsConfigs    []*NotifierConfig[victorops.Config]
+	WebhookConfigs      []*NotifierConfig[webhook.Config]
+	WecomConfigs        []*NotifierConfig[wecom.Config]
+	WebexConfigs        []*NotifierConfig[webex.Config]
+}
+
+// NotifierConfig represents parsed GrafanaReceiver.
+type NotifierConfig[T interface{}] struct {
+	receivers.NotifierInfo
+	Settings T
+}
+
+// ValidateAPIReceiver parses, decrypts and validates the APIReceiver. GrafanaReceiverTyped that contains configurations of all notifiers configurations for this receiver
+func ValidateAPIReceiver(ctx context.Context, api *APIReceiver, decrypt receivers.GetDecryptedValueFn) (GrafanaReceiverTyped, error) {
+	result := GrafanaReceiverTyped{
+		Name: api.Name,
+	}
+	parseConfig := func(receiver *GrafanaReceiver) error {
+		// secure settings are already encrypted at this point
+		secureSettings := make(map[string][]byte, len(receiver.SecureSettings))
+
+		if receiver.SecureSettings != nil {
+			for k, v := range receiver.SecureSettings {
+				d, err := base64.StdEncoding.DecodeString(v)
+				if err != nil {
+					return InvalidReceiverError{
+						Receiver: receiver,
+						Err:      errors.New("failed to decode secure setting"),
+					}
+				}
+				secureSettings[k] = d
+			}
+		}
+
+		var decryptFn receivers.DecryptFunc = func(key string, fallback string) string {
+			return decrypt(ctx, secureSettings, key, fallback)
+		}
+
+		switch strings.ToLower(receiver.Type) {
+		case "prometheus-alertmanager":
+			return createReceiver[alertmanager.Config](receiver)(alertmanager.NewConfig(receiver.Settings, decryptFn))(func(f *NotifierConfig[alertmanager.Config]) {
```

Comment:
I personally find this very hard to read - if all that we're saving is `err != nil` I'd like to move away from this model. 

Discussed this offline, agreed to move away from this model.